File: Examples/Spectrum_Analyzers/src/SignalVuExamples/python/rsa_trace_transfer.py
"""
VISA: RSA Spectrum Trace Transfer
Author: Morgan Allison
Date created: Unknown
Date edited: 5/17
This program transfers the Spectrum trace from the RSA to the 
computer and plots the results.
Windows 7 64-bit, TekVISA 4.0.4
Python 3.6.0 64-bit (Anaconda 4.3.0)
NumPy 1.11.2, MatPlotLib 2.0.0, PyVISA 1.8
To get PyVISA: pip install pyvisa
Download Anaconda: http://continuum.io/downloads
Anaconda includes MatPlotLib
Download SignalVu-PC programmer manual: http://www.tek.com/node/1828803
Download RSA5100B programmer manual: 
http://www.tek.com/spectrum-analyzer/inst5000-manual-7
Tested on RSA306B, RSA507A, and RSA5126B
"""

import visa
import numpy as np
import matplotlib.pyplot as plt

"""#################SEARCH/CONNECT#################"""
rm = visa.ResourceManager()
rsa = rm.open_resource('GPIB8::1::INSTR')
rsa.timeout = 10000
rsa.encoding = 'latin_1'
rsa.write_termination = None
rsa.read_termination = '\n'
print(rsa.query('*idn?'))
rsa.write('*rst')
rsa.write('*cls')
rsa.write('abort')

"""#################CONFIGURE INSTRUMENT#################"""
# configure acquisition parameters
cf = 2.4453e9
span = 40e6
refLevel = 0
rsa.write('spectrum:frequency:center {}'.format(cf))
rsa.write('spectrum:frequency:span {}'.format(span))
rsa.write('input:rlevel {}'.format(refLevel))

rsa.write('initiate:continuous off')
rsa.write('trigger:status off')

"""#################ACQUIRE/PROCESS DATA#################"""
# start acquisition THIS MUST BE DONE
# it is an overlapping command, so *OPC? MUST be sent for synchronization
rsa.write('initiate:immediate')
rsa.query('*opc?')

spectrum = rsa.query_binary_values('fetch:spectrum:trace?', datatype='f',
                                   container=np.array)

# generate the frequency vector for plotting
fMin = cf - span / 2
fMax = cf + span / 2
freq = np.linspace(fMin, fMax, len(spectrum))

"""#################PLOTS#################"""
fig = plt.figure(1, figsize=(15, 8))
ax = fig.add_subplot(111, facecolor='k')
ax.plot(freq / 1e9, spectrum, 'y')
ax.set_title('Spectrum')
ax.set_xlabel('Frequency (GHz)')
ax.set_ylabel('Amplitude (dBm)')
ax.set_xlim(fMin / 1e9, fMax / 1e9)
ax.set_ylim(refLevel - 100, refLevel)
plt.show()

rsa.close()

File: Examples/Spectrum_Analyzers/src/SignalVuExamples/MATLAB/rsa_trace_transfer.m
% VISA Control: RSA Trace Transfer
% Author: Morgan Allison
% Date Edited: 8/17
% This program transfers the Amplitude vs Time trace from the RSA to the 
% computer and plots the results. 
% Windows 7 64-bit, TekVISA 4.0.4
% Matlab r2017a with ICT
% Download SignalVu-PC programmer manual: http://www.tek.com/node/1828803
% Download RSA5100B programmer manual: 
% http://www.tek.com/spectrum-analyzer/inst5000-manual-7
% Download SignalVu programmer manual: 
% http://www.tek.com/oscilloscope/dpo70000-mso70000-manual-22
% Tested on RSA306B/RSA507A with SignalVu-PC 3.10.0030, 
% DPO77002SX with SignalVu 3.9.0051


%% #################SEARCH/CONNECT#################
visaBrand = 'tek';
% instAddress = 'TCPIP::192.168.1.10::INSTR';
instAddress = 'GPIB8::1::INSTR';
buf = 50000;
inst = visa(visaBrand, instAddress);
set(inst, 'InputBuffer', buf);
set(inst, 'OutputBuffer', buf);
fopen(inst);
inst.timeout = 15;

instID = query(inst,'*idn?');
fprintf('Connected to %s\n',instID);

if contains(instID, 'MSO') || contains(instID, 'DPO')
    % IMPORTANT: make sure SignalVu is already running if you're using a scope.
    % The application:activate command gives focus to SignalVu.
    % *OPC? does not respond to application:activate, so there's no good 
    % way to synchronize this command
    sampleRate = 50e9;

    disp('Activating SignalVu');
    fprintf(inst, 'application:activate "SignalVu Vector Signal Analysis Software"');
    fprintf(inst, 'sense:signalvu:acquisition:control:sample:rate off');
    fprintf(inst, 'sense:signalvu:acquisition:digitizer:sample:rate %d', sampleRate);
end

% preset, clear buffer, and stop acquisition
fprintf(inst, 'system:preset');
fprintf(inst, '*cls');
fprintf(inst, 'abort');


%% #################CONFIGURE INSTRUMENT#################%% 
% configure acquisition parameters
cf = 2.4453e9;
span = 40e6;
refLevel = 0;

fprintf(inst, 'spectrum:frequency:center %d', cf);
fprintf(inst, 'spectrum:frequency:span %d', span);
fprintf(inst, 'input:rlevel %d', refLevel);

fprintf(inst, 'initiate:continuous off');
fprintf(inst, 'trigger:status off');

%% #################ACQUIRE/PROCESS DATA#################%% 
% start acquisition THIS MUST BE DONE
% it is an overlapping command, so *OPC? MUST be sent for synchronization
fprintf(inst, 'initiate:immediate');
query(inst, '*opc?');

fprintf(inst, 'fetch:spectrum:trace?');
spectrum = binblockread(inst, 'float');

% generate the frequency vector for plotting
fMin = cf - span / 2;
fMax = cf + span / 2;
freq = linspace(fMin, fMax, length(spectrum));

%% #################PLOTS#################%% 
figure(1);
whitebg(1,'k');
plot((freq/1e9), spectrum, 'y')
title('Spectrum')
xlabel('Frequency (GHz)')
ylabel('Amplitude (dBm)')
xlim([fMin / 1e9, fMax / 1e9])
ylim([refLevel - 100, refLevel])

%% Close instrument
fclose(inst);
delete(inst);
clear inst;